centrallix/expression/exp_functions.c

```diff
@@ -3978,368 +4151,232 @@ int exp_fn_nth(pExpression tree, pParamObjects objlist, pExpression i0, pExpress
     return 0;
     }
 
-/* See centrallix-sysdoc/string_comparison.md for more information. */
-int exp_fn_levenshtein(pExpression tree, pParamObjects objlist, pExpression i0, pExpression i1, pExpression i2)
+static int exp_fn_verify_schema(
+    const char* fn_name,
+    const int* param_types,
+    const int num_params,
+    pExpression tree,
+    pParamObjects obj_list)
     {
-
-    if (!i0 || !i1)
+    /** Verify object list and session. **/
+    if (obj_list == NULL)
 	{
-		mssError(1,"EXP","levenshtein() requires two parameters");
-		return -1;
+	mssErrorf(1, "EXP", "%s(\?\?\?) no object list?", fn_name);
+	return -1;
 	}
-
-    if ((i0->Flags & EXPR_F_NULL) || (i1->Flags & EXPR_F_NULL))
-	{
-		tree->DataType = DATA_T_INTEGER;
-		tree->Flags |= EXPR_F_NULL;
-		return 0;
+    ASSERTMAGIC(obj_list->Session, MGK_OBJSESSION);
+    
+    /** Verify expression tree. **/
+    ASSERTMAGIC(tree, MGK_EXPRESSION);
+    
+    /** Verify parameter number. **/
+    const int num_params_actual = tree->Children.nItems;
+    if (num_params != num_params_actual)
+	{
+	mssErrorf(1, "EXP",
+	    "%s(?) expects %u param%s, got %d param%s.",
+	    fn_name, num_params, (num_params > 1) ? "s" : "", num_params_actual, (num_params_actual > 1) ? "s" : ""
+	);
+	return -1;
 	}
-
-    if ((i0->DataType != DATA_T_STRING) || (i1->DataType != DATA_T_STRING))
+        
+    /** Verify parameter datatypes. **/
+    for (int i = 0; i < num_params; i++)
 	{
-		mssError(1,"EXP","levenshtein() requires two string parameters");
-		return -1;
+	const pExpression arg = tree->Children.Items[i];
+	ASSERTMAGIC(arg, MGK_EXPRESSION);
+	
+	/** Skip null values. **/
+	if (arg->Flags & EXPR_F_NULL) continue;
+	
+	/** Extract datatypes. **/
```

Comment:
We'll need some kind of option for functions accepting more than one data type (for example, integer, double, or money).  One option is to use DATA_T_ANY as a datatype and have it match anything, and leave it up to the individual function implementation to validate that data type if not all types are acceptable.

Reply:
I merged the schema verification branch. The new function is able to handle this.